src/lib/Repository/NameSchema/NameSchemaService.php

```diff
@@ -0,0 +1,423 @@
+                        $groupToken,
+                        $this->resolveToken(
+                            $groupToken,
+                            $titles,
+                            $groupLookupTable
+                        ),
+                        $replaceString
+                    );
+                }
+
+                // We want to stop after the first matching token part / identifier is found
+                // <id1|id2> if id1 has a value, id2 will not be used.
+                // In this case id1 or id1 is a token group.
+                break;
+            } else {
+                if (array_key_exists(
+                    $tokenPart,
+                    $titles
+                ) && $titles[$tokenPart] !== '' && $titles[$tokenPart] !== null) {
+                    $replaceString = $titles[$tokenPart];
+                    // We want to stop after the first matching token part / identifier is found
+                    // <id1|id2> if id1 has a value, id2 will not be used.
+                    break;
+                }
+            }
+        }
+
+        return $replaceString;
+    }
+
+    /**
+     * Checks whether $identifier is a placeholder for a token group.
+     *
+     * @param string $identifier
+     *
+     * @return bool
+     */
+    protected function isTokenGroup(string $identifier): bool
+    {
+        if (strpos($identifier, self::META_STRING) === false) {
+            return false;
+        }
+
+        return true;
+    }
+
+    /**
+     * Returns the different constituents of $token in an array.
+     * The normal case here is that the different identifiers within one token
+     * will be tokenized and returned.
+     *
+     * Example:
+     * <code>
+     * "&lt;title|text&gt;" ==&gt; array( 'title', 'text' )
+     * </code>
+     *
+     * @param string $token
+     *
+     * @return array
+     */
+    protected function tokenParts(string $token): array
+    {
+        return preg_split('/[^\w:]+/', $token, -1, PREG_SPLIT_NO_EMPTY);
+    }
+
+    /**
+     * Builds a lookup / translation table for groups in the $namePattern.
+     * The groups are referenced with a generated meta-token in the original
+     * name pattern.
+     *
+     * Returns intermediate name pattern where groups are replaced with meta-
+     * tokens.
+     *
+     * @param string $nameSchema
+     *
+     * @return array
+     */
+    protected function filterNameSchema(string $nameSchema): array
+    {
+        $retNamePattern = '';
+        $foundGroups = preg_match_all('/\((.+)\)/U', $nameSchema, $groupArray);
+        $groupLookupTable = [];
+
+        if ($foundGroups) {
+            $i = 0;
+            foreach ($groupArray[1] as $group) {
+                // Create meta-token for group
+                $metaToken = self::META_STRING . $i;
+
+                // Insert the group with its placeholder token
+                $retNamePattern = str_replace($group, $metaToken, $nameSchema);
```

Comment:
We need to help a bit static analysis getting confused by crappy PHP functions design.
```suggestion
                $retNamePattern = str_replace($group, $metaToken, $nameSchema);
                /** @var string $retNamePattern */
```